python_ta/checkers/static_type_checker.py

```diff
@@ -0,0 +1,287 @@
+"""
+A checker that performs static type checking using Mypy.
+This module integrates Mypy to detect issues related to type annotations in Python code
+"""
+
+import re
+from typing import Optional
+
+from astroid import nodes
+from mypy import api
+from pylint.checkers import BaseRawFileChecker
+from pylint.checkers.utils import only_required_for_messages
+from pylint.lint import PyLinter
+
+
+class StaticTypeChecker(BaseRawFileChecker):
+    """Checker for static type checking using Mypy."""
+
+    name = "static_type_checker"
+    msgs = {
+        "E9951": (
+            "Argument %s to %s has incompatible type %s; expected %s",
+            "incompatible-argument-type",
+            "Used when a function argument has an incompatible type",
+        ),
+        "E9952": (
+            "Incompatible types in assignment (expression has type %s, variable has type %s)",
+            "incompatible-assignment",
+            "Used when there is an incompatible assignment",
+        ),
+        "E9953": (
+            "List item %s has incompatible type %s; expected %s",
+            "list-item-type-mismatch",
+            "Used when a list item has an incompatible type",
+        ),
+        "E9954": (
+            "Unsupported operand types for %s (%s and %s)",
+            "unsupported-operand-types",
+            "Used when an operation is attempted between incompatible types",
+        ),
+        "E9955": (
+            "Item of type %s in Union has no attribute %s",
+            "union-attr-error",
+            "Used when accessing an attribute that may not exist on a Union type",
+        ),
+        "E9956": (
+            "Dict entry %s has incompatible type %s: %s; expected %s: %s",
+            "dict-item-type-mismatch",
+            "Used when a dictionary entry has an incompatible key or value type",
+        ),
+    }
+
+    @only_required_for_messages(
+        "incompatible-argument-type",
+        "incompatible-assignment",
+        "list-item-type-mismatch",
+        "unsupported-operand-types",
+        "union-attr-error",
+        "dict-item-type-mismatch",
+    )
+    def process_module(self, node: nodes.NodeNG) -> None:
+        """Run Mypy on the current file and handle type errors."""
+        filename = node.stream().name
+        mypy_options = [
+            "--ignore-missing-imports",
+            "--disable-error-code=call-arg",
+            "--show-error-end",
+        ]
+        result, _, _ = api.run([filename] + mypy_options)
+        for line in result.splitlines():
+            if line.endswith("[arg-type]"):
```

Comment:
Overall you can simplify this code as follows:

1. Use one regular expression to extract the *common elements* from each message: the file name, line/col coordinates, error message, and error code. This is a consistent format for all messages.
2. Maintain a dictionary mapping error code (e.g. `'incompatible-argument-type'`) to regular expression that extracts the relevant parts of the message (e.g., `arg_num`, `func_name`, `incomp_type`, `exp_type`).
3. Then you can have just a single block of code here that first parses the common parts, then (if successful) parses the message-specific parts, and then calls `self._add_message`.